segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/rolling/RollingStorage.java

```diff
@@ -0,0 +1,747 @@
+                try {
+                    log.warn("Could not create Header Segment for '{}', rolling back.", segmentName, ex);
+                    this.baseStorage.delete(headerHandle);
+                } catch (Exception ex2) {
+                    ex.addSuppressed(ex2);
+                }
+            }
+
+            throw ex;
+        }
+
+        LoggerHelpers.traceLeave(log, "create", traceId, segmentName);
+        return StreamSegmentInformation.builder().name(segmentName).build();
+    }
+
+    @Override
+    public SegmentHandle openWrite(String segmentName) throws StreamSegmentException {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", segmentName);
+        val handle = openHandle(segmentName, false);
+
+        // Finally, open the Active SegmentChunk for writing.
+        SegmentChunk last = handle.lastChunk();
+        if (last != null && !last.isSealed()) {
+            val activeHandle = this.baseStorage.openWrite(last.getName());
+            handle.setActiveChunkHandle(activeHandle);
+        }
+
+        LoggerHelpers.traceLeave(log, "openWrite", traceId, handle);
+        return handle;
+    }
+
+    @Override
+    public void write(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentException {
+        val h = asWritableHandle(handle);
+        ensureNotDeleted(h);
+        ensureNotSealed(h);
+        ensureOffset(h, offset);
+        long traceId = LoggerHelpers.traceEnter(log, "write", handle, offset, length);
+
+        // We run this in a loop because we may have to split the write over multiple SegmentChunks in order to avoid exceeding
+        // any SegmentChunk's maximum length.
+        int bytesWritten = 0;
+        while (bytesWritten < length) {
+            if (h.getActiveChunkHandle() == null || h.lastChunk().getLength() >= h.getRollingPolicy().getMaxLength()) {
+                rollover(h);
+            }
+
+            SegmentChunk last = h.lastChunk();
+            int writeLength = (int) Math.min(length - bytesWritten, h.getRollingPolicy().getMaxLength() - last.getLength());
+            assert writeLength > 0 : "non-positive write length";
+            long chunkOffset = offset + bytesWritten - last.getStartOffset();
+            this.baseStorage.write(h.getActiveChunkHandle(), chunkOffset, data, writeLength);
+            last.increaseLength(writeLength);
+            bytesWritten += writeLength;
+        }
+
+        LoggerHelpers.traceLeave(log, "write", traceId, handle, offset, bytesWritten);
+    }
+
+    @Override
+    public void seal(SegmentHandle handle) throws StreamSegmentException {
+        val h = asWritableHandle(handle);
+        ensureNotDeleted(h);
+        long traceId = LoggerHelpers.traceEnter(log, "seal", handle);
+        sealActiveChunk(h);
+        SegmentHandle headerHandle = h.getHeaderHandle();
+        if (headerHandle != null) {
+            this.baseStorage.seal(headerHandle);
```

Comment:
We seal the active chunk as well (see 3 lines above). In case only that one succeeds, and we're not able to seal the header too, then the operation will fail. A subsequent retry will detect that the segment is not actually sealed, and only seal the header file. If the app decides not to retry, and just continue writing, then the `write` operation will simply create a new chunk and write to it.